### PR TITLE
Fix issue #418. Unpin pymbolic. Resolve pydantic deprecation warnings

### LINKED ISSUE #418: Params is_symbolic failing on linux
### Setup Information

- ravenpy version: 0.16.0
- Python version: 3.12.7
- Operating System: Issue on WSL2 and linux cluster. Fine on windows

### Description

In config/rvs.py line 392 asdict(self.params) looses "Variable" : See screenshot below :
dda6ff6d-ce85-4fa5-9809-e158302e1551

So line 520 looking for Variable instances returns False and SpotSetup cannot initialize

On Windows, it's working fine :
945c4461-0040-4340-80cc-8b46f77e1017

### Steps To Reproduce

_No response_

### Additional context

_No response_

### Contribution

- [ ] I would be willing/able to open a Pull Request to address this bug.

### PR DESCRIPTION
**Pull request by huard**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
  - This PR fixes #418 
- [ ] (If applicable) Documentation has been added / updated (for bug fixes / features).
- [ ] (If applicable) Tests have been added.
- [x] CHANGELOG.rst has been updated (with summary of main changes).
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.

### What kind of change does this PR introduce?

* Unpin pymbolic
* Resolve issue with `pymbolic > 2022.2`
* Tweaks to silence pydantic DeprecationWarning

### Does this PR introduce a breaking change?

Python3.9 support has been dropped.

### CODE REVIEW

**Comment by Zeitsperre** [comment]
@huard If you want to implement a drop of Python3.9 in order to address the problem, go for it. We've been progressively dropping that version in a few places already.

File: pyproject.toml [hunk 1]
```diff
@@ -53,7 +52,7 @@ dependencies = [
   "platformdirs",
   "pydantic >=2.0",
   "pydap >=3.4.0",
-  "pymbolic <=2022.2", # 2024.1.0 is not supported
+  "pymbolic",
```
Comment (Zeitsperre) [inline pyproject.toml:55]: It would be good to have the lowest version that the new vocabulary supports.

File: environment-dev.yml [hunk 2]
```diff
@@ -29,7 +29,7 @@ dependencies:
   - pydantic >=2.0
   - pydap >=3.4.0
   - pymetalink >=6.5.2
-  - pymbolic <=2022.2 # 2024.1.0 is not supported
+  - pymbolic
```
Comment (Zeitsperre) [inline environment-dev.yml:32]: Having a base version would be good here.

File: src/ravenpy/config/base.py [hunk 3]
```diff
@@ -2,10 +2,10 @@
 from collections.abc import Sequence
 from enum import Enum
 from textwrap import dedent, indent
-from typing import Any, Optional, Union
+from typing import Any, Optional, TypedDict, Union
```
Comment (Zeitsperre) [inline src/ravenpy/config/base.py:5]:
```suggestion
from typing import Any, Optional, Union
```

**Review by Zeitsperre — APPROVED** [review]
LGTM!

A few final changes:
- In `.pre-commit.config.yml`, change the `pyupgrade` hook to `py310-plus`
- In `pyproject.toml`, change the `ruff` and `isort` configs to use `"310"` instead of `"39"`, and drop `py39` from `black`. 

The remaining formatting should be automated by pre-commit.

File: pyproject.toml (hunk 1 above)
Comment (huard) [reply pyproject.toml:55]: Agreed, but unclear from a quick look at the release notes. Didn't really have time to hunt the info down. What do you suggest?

Comment (Zeitsperre) [reply pyproject.toml:55]: I need to fix up the formatting, so I'll see what I can do this morning.

**Pull request merged**